### ISSUE TITLE
Fanning sums

### ISSUE THREAD

**Issue by treeowl** [reflex-frp/reflex#316]
I may not understand this properly, but at first glance it seems that `fanEither` should generalize to something like

```haskell
fanSum
  :: (GCompare tag, GEnum tag)
  => Event t (DSum tag f) -> DMap tag (Compose (Event t) f)
```

where `GEnum` offers a way to enumerate all the tags. (Possible methods: `enumTags :: [DSum tag Proxy]`, `enumTagsMap :: DMap tag Proxy`).

Simplifying to `Identity`, that's

```haskell
fanSumId
  :: (GCompare tag, GEnum tag)
  => Event t (DSum tag Identity) -> DMap tag (Event t)
```

The idea is that if we have a stream of messages of various types, we should be able to separate it into multiple streams, each of which contains messages of only one type. The big question is whether there's a way to implement this efficiently. I don't have a clue myself.

Note: I've opened [an issue](https://github.com/mokus0/dependent-sum/issues/29) for `dependent-sum` suggesting such a class.

**Comment by ryantrinkle** [comment]
Have you looked at the `fan` primitive? I think that may do what
you're looking for.

**Comment by 3noch** [comment]
Have you looked at `factorEvent`?

**Comment by treeowl** [comment]
@ryantrinkle Ah, I think I see what you mean. Yes, that's probably sufficient, but it seems like overkill. I can turn each `DSum` into a singleton `DMap` and use `fan` with that, but it seems clearer (and at least a bit more efficient) to avoid that.

**Comment by treeowl** [comment]
@3noch `factorEvent` doesn't look likely, since it's all about paying attention to when the key changes and I don't care about that.

**Comment by ryantrinkle** [comment]
Possibly, but the hard work is mostly on the subscriber management side,
and it wouldn't be reduced by inputting DSums rather than DMaps. That being
said, this all sounds like interesting stuff and I look forward to seeing
how it goes!

**Comment by treeowl** [comment]
If you don't think it matters, then that's fine. It would at least be nice to offer it as a function if not a method, since it takes a bit of mental gymnastics to see that `fan` does the job.

**Comment by treeowl** [comment]
Oh, but there is still the matter of type .... Why is `fan` only for `Identity`?

**Comment by ryantrinkle** [comment]
Oh, that could probably be changed. I think it's just a holdover from when
DMap was one-argument

**Comment by saulzar** [comment]
Implemented as PR #318
